relay-quotas/src/global.rs

```diff
@@ -0,0 +1,231 @@
+/// Calling Redis every time we want to check if an item should be ratelimited would be very expensive,
+/// which is why we have this cache. It works by 'taking' a certain budget from Redis, by pre-incrementing
+/// a global counter. We Put the amount we pre-incremented into this local cache and count down until
+/// we have no more budget, then we ask for more from Redis. If we find the global counter is above
+/// the quota limit, we will ratelimit the item.
+struct GlobalRateLimit {
+    local_counter: LocalCounter,
+    redis_counter: RedisCounter,
+    slot: usize,
+}
+
+impl GlobalRateLimit {
+    pub fn new() -> Self {
+        Self {
+            local_counter: LocalCounter::new(),
+            redis_counter: RedisCounter::new(),
+            slot: 0,
+        }
+    }
+
+    pub fn is_rate_limited(
+        &mut self,
+        client: &mut PooledClient,
+        quota: &RedisQuota,
+        quantity: usize,
+        limit: u64,
+    ) -> Result<bool, RedisError> {
+        let current_slot = current_slot(quota.window());
+
+        if current_slot < self.slot {
+            // TODO double check logic here
+            // in theory this should never happen only if someone messes with the system time
+            // be safe and dont rate limit
+            relay_log::error!("budget slot ahead of current slot");
+            return Ok(false);
+        }
+
+        if current_slot > self.slot {
+            // We moved to the next slot and need to reset the local counter,
+            // Redis already handles this by having the slot in the key.
+            self.local_counter = LocalCounter::new();
+            self.slot = current_slot;
+        }
+
+        if self.local_counter.try_consume(quantity) {
+            return Ok(false);
+        }
+
+        if !self.redis_counter.can_satisfy(quantity, limit) {
+            return Ok(true);
+        }
+
+        let budget_to_reserve = quantity.max(self.default_request_size_based_on_limit());
+        let reserved = self
+            .redis_counter
+            .try_reserve(client, budget_to_reserve, limit, quota)? as usize;
+
+        self.local_counter.increase_budget(reserved);
+        Ok(!self.local_counter.try_consume(quantity))
+    }
+
+    fn default_request_size_based_on_limit(&self) -> usize {
+        100
+    }
+}
+
+impl Default for GlobalRateLimit {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+struct LocalCounter {
+    budget: usize,
+}
+
+impl LocalCounter {
+    fn new() -> Self {
+        Self { budget: 0 }
+    }
+
+    fn try_consume(&mut self, quantity: usize) -> bool {
+        if self.budget >= quantity {
+            self.budget -= quantity;
+            true
+        } else {
+            false
+        }
+    }
+
+    fn increase_budget(&mut self, quantity: usize) {
+        self.budget = self.budget.saturating_add(quantity);
+    }
+}
+
+struct RedisCounter {
+    last_seen: u64,
```

Comment:
nit: Might be just me but I associate `last_seen` with a timestamp.
```suggestion
    latest: u64,
```

Reply:
We prefer `last_seen` because it makes it clear the value may be out of date and not in sync with Redis.